src/cargo/util/errors.rs

```diff
@@ -132,6 +132,50 @@ impl fmt::Debug for ProcessError {
 }
 
 // =============================================================================
+// Cargo test errors.
+
+/// Error when testcases fail
+pub struct CargoTestError {
+    pub desc: String,
+    pub exit: Option<ExitStatus>,
+    cause: Option<io::Error>,
+}
+
+impl fmt::Display for CargoTestError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(&self.desc, f)
+    }
+}
+
+impl fmt::Debug for CargoTestError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+impl Error for CargoTestError {
+    fn description(&self) -> &str { &self.desc }
+    #[allow(trivial_casts)]
+    fn cause(&self) -> Option<&Error> {
+        self.cause.as_ref().map(|s| s as &Error)
+    }
+}
+
+#[allow(deprecated)] // connect => join in 1.3
+impl<'a> From<&'a [ProcessError]> for CargoTestError {
```

Comment:
This may actually be best expressed as just a `new` constructor because then the concrete type of the argument is known so you don't need to do `&foo[..]` but instead can just do `&foo`